File: packages/face-detection/ios/HybridFaceDetector.swift
import Foundation
import UIKit
import NitroModules
import MLKitFaceDetection
import MLKitVision

/**
 * Native iOS implementation of FaceDetector.
 * Conforms to the Nitrogen-generated HybridFaceDetectorSpec protocol.
 */
class HybridFaceDetector: HybridFaceDetectorSpec {
    
    // MARK: - HybridObject boilerplate
    var hybridContext = margelo.nitro.HybridContext()
    var memorySize: Int { MemoryLayout<HybridFaceDetector>.size }
    
    // MARK: - Lazy MLKit detectors
    
    private lazy var fastDetector: MLKitFaceDetection.FaceDetector = {
        let opts = FaceDetectorOptions()
        opts.performanceMode = .fast
        opts.landmarkMode = .none
        opts.classificationMode = .none
        return MLKitFaceDetection.FaceDetector.faceDetector(options: opts)
    }()
    
    private lazy var accurateDetector: MLKitFaceDetection.FaceDetector = {
        let opts = FaceDetectorOptions()
        opts.performanceMode = .accurate
        opts.landmarkMode = .all
        opts.classificationMode = .all
        return MLKitFaceDetection.FaceDetector.faceDetector(options: opts)
    }()
    
    // MARK: - Protocol methods
    
    func detect(imageUri: String, options: FaceDetectionOptions) throws -> Promise<[DetectedFace]> {
        return Promise.async {
            guard let image = self.loadImage(from: imageUri) else {
                throw RuntimeError.error(withMessage: "Failed to load image: \(imageUri)")
            }
            let visionImage = VisionImage(image: image)
            visionImage.orientation = image.imageOrientation
            
            let detector = options.performanceMode == .accurate ? self.accurateDetector : self.fastDetector
            let mlFaces = try await detector.process(visionImage)
            
            return mlFaces.map { self.mapFace($0, withLandmarks: options.landmarks) }
        }
    }
    
    func detectBatch(imageUris: [String], concurrency: Double) throws -> Promise<[BatchCropResult]> {
        return Promise.async {
            let maxConcurrent = Int(concurrency)
            var results = [BatchCropResult]()
            results.reserveCapacity(imageUris.count)
            
            // Process in chunks for controlled concurrency
            for chunkStart in stride(from: 0, to: imageUris.count, by: maxConcurrent) {
                let chunkEnd = min(chunkStart + maxConcurrent, imageUris.count)
                let chunk = Array(imageUris[chunkStart..<chunkEnd])
                
                let chunkResults = await withTaskGroup(of: (Int, BatchCropResult).self) { group in
                    for (i, uri) in chunk.enumerated() {
                        let globalIdx = chunkStart + i
                        group.addTask {
                            do {
                                guard let img = self.loadImage(from: uri) else {
                                    return (globalIdx, BatchCropResult(index: Double(globalIdx), faces: [], crops: [], success: false))
                                }
                                let visionImage = VisionImage(image: img)
                                let mlFaces = try await self.fastDetector.process(visionImage)
                                let faces = mlFaces.map { self.mapFace($0, withLandmarks: false) }
                                return (globalIdx, BatchCropResult(index: Double(globalIdx), faces: faces, crops: [], success: true))
                            } catch {
                                return (globalIdx, BatchCropResult(index: Double(globalIdx), faces: [], crops: [], success: false))
                            }
                        }
                    }
                    
                    var chunkOut = [(Int, BatchCropResult)]()
                    for await result in group { chunkOut.append(result) }
                    return chunkOut.sorted(by: { $0.0 < $1.0 }).map(\.1)
                }
                results.append(contentsOf: chunkResults)
            }
            return results
        }
    }
    
    func detectPrimary(imageUri: String) throws -> Promise<DetectedFace> {
        return Promise.async {
            guard let image = self.loadImage(from: imageUri) else {
                throw RuntimeError.error(withMessage: "Failed to load image: \(imageUri)")
            }
            let visionImage = VisionImage(image: image)
            let mlFaces = try await self.accurateDetector.process(visionImage)
            
            guard let largest = mlFaces.max(by: {
                $0.frame.width * $0.frame.height < $1.frame.width * $1.frame.height
            }) else {
                throw RuntimeError.error(withMessage: "No face detected")
            }
            return self.mapFace(largest, withLandmarks: true)
        }
    }
    
    func cropFaces(imageUri: String, padding: Double) throws -> Promise<[CroppedFace]> {
        return Promise.async {
            guard let image = self.loadImage(from: imageUri),
                  let cgImage = image.cgImage else {
                throw RuntimeError.error(withMessage: "Failed to load image")
            }
            let visionImage = VisionImage(image: image)
            let mlFaces = try await self.fastDetector.process(visionImage)
            
            let imgW = Double(cgImage.width)
            let imgH = Double(cgImage.height)
            var crops = [CroppedFace]()
            
            for (idx, face) in mlFaces.enumerated() {
                let bounds = face.frame
                let padX = Double(bounds.width) * padding
                let padY = Double(bounds.height) * padding
                
                let x = max(0, Double(bounds.origin.x) - padX)
                let y = max(0, Double(bounds.origin.y) - padY)
                let w = min(imgW - x, Double(bounds.width) + padX * 2)
                let h = min(imgH - y, Double(bounds.height) + padY * 2)
                
                let cropRect = CGRect(x: x, y: y, width: w, height: h)
                guard let cropped = cgImage.cropping(to: cropRect) else { continue }
                
                let tempPath = NSTemporaryDirectory() + "nitro_face_\(UUID().uuidString).jpg"
                let croppedImg = UIImage(cgImage: cropped)
                guard let data = croppedImg.jpegData(compressionQuality: 0.8) else { continue }
                try data.write(to: URL(fileURLWithPath: tempPath))
                
                crops.append(CroppedFace(
                    uri: "file://\(tempPath)",
                    faceIndex: Double(idx),
                    width: w,
                    height: h
                ))
            }
            return crops
        }
    }
    
    func extractEmbedding(faceUri: String) throws -> Promise<[Double]> {
        return Promise.async {
            // TODO: Load MobileFaceNet TFLite model and run inference
            // For now return placeholder embedding
            throw RuntimeError.error(withMessage: "MobileFaceNet model not yet loaded. Coming in v0.2.0")
        }
    }
    
    func compareFaces(embedding1: [Double], embedding2: [Double]) throws -> Double {
        // Cosine similarity
        guard embedding1.count == embedding2.count, !embedding1.isEmpty else { return 0 }
        var dot = 0.0, mag1 = 0.0, mag2 = 0.0
        for i in 0..<embedding1.count {
            dot += embedding1[i] * embedding2[i]
            mag1 += embedding1[i] * embedding1[i]
            mag2 += embedding2[i] * embedding2[i]
        }
        let denom = sqrt(mag1) * sqrt(mag2)
        return denom > 0 ? (dot / denom + 1.0) / 2.0 : 0 // normalize to 0..1
    }
    
    func extractPrimaryEmbedding(imageUri: String) throws -> Promise<[Double]> {
        return Promise.async {
            // TODO: detect primary face → crop → embed with MobileFaceNet
            throw RuntimeError.error(withMessage: "MobileFaceNet model not yet loaded. Coming in v0.2.0")
        }
    }
    
    func detectAndEmbed(imageUris: [String], concurrency: Double) throws -> Promise<[BatchEmbeddingResult]> {
        return Promise.async {
            // TODO: batch detect + embed
            throw RuntimeError.error(withMessage: "MobileFaceNet model not yet loaded. Coming in v0.2.0")
        }
    }
    
    func isAvailable() throws -> Bool {
        return true
    }
    
    // MARK: - Private helpers
    
    private func loadImage(from uri: String) -> UIImage? {
        let path: String
        if uri.hasPrefix("file://") {
            path = String(uri.dropFirst(7))
        } else if uri.hasPrefix("/") {
            path = uri
        } else {
            guard let url = URL(string: uri), let data = try? Data(contentsOf: url) else { return nil }
            return UIImage(data: data)
        }
        return UIImage(contentsOfFile: path)
    }
    
    private func mapFace(_ face: Face, withLandmarks: Bool) -> DetectedFace {
        var landmarks = [FaceLandmark]()
        if withLandmarks {
            let pairs: [(FaceLandmarkType, MLKitFaceDetection.FaceLandmarkType)] = [
                (.leftEye, .leftEye), (.rightEye, .rightEye),
                (.leftEar, .leftEar), (.rightEar, .rightEar),
                (.leftCheek, .leftCheek), (.rightCheek, .rightCheek),
                (.noseBase, .noseBase), (.mouthLeft, .mouthLeft),
                (.mouthRight, .mouthRight), (.mouthBottom, .mouthBottom),
            ]
            for (type, mlType) in pairs {
                if let lm = face.landmark(ofType: mlType) {
                    landmarks.append(FaceLandmark(type: type, x: Double(lm.position.x), y: Double(lm.position.y)))
                }
            }
        }
        
        return DetectedFace(
            bounds: FaceBounds(
                x: Double(face.frame.origin.x),
                y: Double(face.frame.origin.y),
                width: Double(face.frame.width),
                height: Double(face.frame.height)
            ),
            headEulerAngleY: Double(face.headEulerAngleY),
            headEulerAngleZ: Double(face.headEulerAngleZ),
            leftEyeOpenProbability: Double(face.leftEyeOpenProbability),
            rightEyeOpenProbability: Double(face.rightEyeOpenProbability),
            smilingProbability: Double(face.smilingProbability),
            landmarks: landmarks,
            trackingId: Double(face.hasTrackingID ? face.trackingID : -1)
        )
    }
}

File: packages/face-detection/android/src/main/kotlin/com/margelo/nitro/nitromlkit/face/HybridFaceDetector.kt
package com.margelo.nitro.nitromlkit.face

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.tasks.await
import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.util.UUID
import com.google.mlkit.vision.face.FaceLandmark as MlkitFaceLandmark

/**
 * Native Android implementation of FaceDetector.
 * Extends the Nitrogen-generated HybridFaceDetectorSpec abstract class.
 */
class HybridFaceDetector : HybridFaceDetectorSpec() {

  // MARK: - Lazy MLKit detectors
  // Mirrors the iOS implementation: two fixed detector configs selected by
  // performanceMode. minFaceSize/tracking options are not wired up yet (same
  // limitation as iOS) — coming with a dynamic detector config in a follow-up.

  private val fastDetector by lazy {
    FaceDetection.getClient(
      FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
        .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
        .build()
    )
  }

  private val accurateDetector by lazy {
    FaceDetection.getClient(
      FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
        .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
        .build()
    )
  }

  // MARK: - Spec methods

  override fun detect(imageUri: String, options: FaceDetectionOptions): Promise<Array<DetectedFace>> {
    return Promise.async {
      val bitmap = loadBitmap(imageUri)
        ?: throw Error("Failed to load image: $imageUri")
      val image = InputImage.fromBitmap(bitmap, 0)

      val detector = if (options.performanceMode == PerformanceMode.ACCURATE) accurateDetector else fastDetector
      val mlFaces = detector.process(image).await()

      mlFaces.map { mapFace(it, withLandmarks = options.landmarks) }.toTypedArray()
    }
  }

  override fun detectBatch(imageUris: Array<String>, concurrency: Double): Promise<Array<BatchCropResult>> {
    return Promise.async {
      val maxConcurrent = concurrency.toInt().coerceAtLeast(1)
      val results = mutableListOf<BatchCropResult>()

      var chunkStart = 0
      while (chunkStart < imageUris.size) {
        val chunkEnd = minOf(chunkStart + maxConcurrent, imageUris.size)
        val chunkResults = coroutineScope {
          (chunkStart until chunkEnd).map { idx ->
            async { detectForBatch(idx, imageUris[idx]) }
          }.awaitAll()
        }
        results.addAll(chunkResults)
        chunkStart = chunkEnd
      }
      results.toTypedArray()
    }
  }

  override fun detectPrimary(imageUri: String): Promise<DetectedFace> {
    return Promise.async {
      val bitmap = loadBitmap(imageUri)
        ?: throw Error("Failed to load image: $imageUri")
      val image = InputImage.fromBitmap(bitmap, 0)
      val mlFaces = accurateDetector.process(image).await()

      val largest = mlFaces.maxByOrNull { it.boundingBox.width().toLong() * it.boundingBox.height().toLong() }
        ?: throw Error("No face detected")
      mapFace(largest, withLandmarks = true)
    }
  }

  override fun cropFaces(imageUri: String, padding: Double): Promise<Array<CroppedFace>> {
    return Promise.async {
      val bitmap = loadBitmap(imageUri)
        ?: throw Error("Failed to load image: $imageUri")
      val image = InputImage.fromBitmap(bitmap, 0)
      val mlFaces = fastDetector.process(image).await()

      val imgW = bitmap.width
      val imgH = bitmap.height
      val context = NitroModules.applicationContext
        ?: throw Error("No application context available")
      val crops = mutableListOf<CroppedFace>()

      mlFaces.forEachIndexed { idx, face ->
        val bounds = face.boundingBox
        val padX = (bounds.width() * padding).toInt()
        val padY = (bounds.height() * padding).toInt()

        val x = maxOf(0, bounds.left - padX)
        val y = maxOf(0, bounds.top - padY)
        val w = minOf(imgW - x, bounds.width() + padX * 2)
        val h = minOf(imgH - y, bounds.height() + padY * 2)
        if (w <= 0 || h <= 0) return@forEachIndexed

        val cropped = Bitmap.createBitmap(bitmap, x, y, w, h)
        val tempFile = File(context.cacheDir, "nitro_face_${UUID.randomUUID()}.jpg")
        FileOutputStream(tempFile).use { out ->
          cropped.compress(Bitmap.CompressFormat.JPEG, 80, out)
        }

        crops.add(
          CroppedFace(
            uri = "file://${tempFile.absolutePath}",
            faceIndex = idx.toDouble(),
            width = w.toDouble(),
            height = h.toDouble()
          )
        )
      }
      crops.toTypedArray()
    }
  }

  override fun extractEmbedding(faceUri: String): Promise<DoubleArray> {
    return Promise.async {
      // TODO: Load MobileFaceNet TFLite model and run inference
      throw Error("MobileFaceNet model not yet loaded. Coming in v0.2.0")
    }
  }

  override fun compareFaces(embedding1: DoubleArray, embedding2: DoubleArray): Double {
    // Cosine similarity, normalized to 0..1
    if (embedding1.size != embedding2.size || embedding1.isEmpty()) return 0.0
    var dot = 0.0
    var mag1 = 0.0
    var mag2 = 0.0
    for (i in embedding1.indices) {
      dot += embedding1[i] * embedding2[i]
      mag1 += embedding1[i] * embedding1[i]
      mag2 += embedding2[i] * embedding2[i]
    }
    val denom = kotlin.math.sqrt(mag1) * kotlin.math.sqrt(mag2)
    return if (denom > 0) (dot / denom + 1.0) / 2.0 else 0.0
  }

  override fun extractPrimaryEmbedding(imageUri: String): Promise<DoubleArray> {
    return Promise.async {
      // TODO: detect primary face -> crop -> embed with MobileFaceNet
      throw Error("MobileFaceNet model not yet loaded. Coming in v0.2.0")
    }
  }

  override fun detectAndEmbed(imageUris: Array<String>, concurrency: Double): Promise<Array<BatchEmbeddingResult>> {
    return Promise.async {
      // TODO: batch detect + embed
      throw Error("MobileFaceNet model not yet loaded. Coming in v0.2.0")
    }
  }

  override fun isAvailable(): Boolean {
    return true
  }

  // MARK: - Private helpers

  private suspend fun detectForBatch(idx: Int, uri: String): BatchCropResult {
    return try {
      val bitmap = loadBitmap(uri)
        ?: return BatchCropResult(idx.toDouble(), emptyArray(), emptyArray(), false)
      val image = InputImage.fromBitmap(bitmap, 0)
      val mlFaces = fastDetector.process(image).await()
      val faces = mlFaces.map { mapFace(it, withLandmarks = false) }.toTypedArray()
      BatchCropResult(idx.toDouble(), faces, emptyArray(), true)
    } catch (e: Exception) {
      BatchCropResult(idx.toDouble(), emptyArray(), emptyArray(), false)
    }
  }

  private fun loadBitmap(uri: String): Bitmap? {
    return try {
      when {
        uri.startsWith("file://") -> BitmapFactory.decodeFile(uri.removePrefix("file://"))
        uri.startsWith("/") -> BitmapFactory.decodeFile(uri)
        uri.startsWith("content://") -> {
          val context = NitroModules.applicationContext ?: return null
          context.contentResolver.openInputStream(Uri.parse(uri))?.use { BitmapFactory.decodeStream(it) }
        }
        else -> URL(uri).openStream().use { BitmapFactory.decodeStream(it) }
      }
    } catch (e: Exception) {
      null
    }
  }

  private fun mapFace(face: Face, withLandmarks: Boolean): DetectedFace {
    val landmarks = if (withLandmarks) {
      val pairs = listOf(
        FaceLandmarkType.LEFT_EYE to MlkitFaceLandmark.LEFT_EYE,
        FaceLandmarkType.RIGHT_EYE to MlkitFaceLandmark.RIGHT_EYE,
        FaceLandmarkType.LEFT_EAR to MlkitFaceLandmark.LEFT_EAR,
        FaceLandmarkType.RIGHT_EAR to MlkitFaceLandmark.RIGHT_EAR,
        FaceLandmarkType.LEFT_CHEEK to MlkitFaceLandmark.LEFT_CHEEK,
        FaceLandmarkType.RIGHT_CHEEK to MlkitFaceLandmark.RIGHT_CHEEK,
        FaceLandmarkType.NOSE_BASE to MlkitFaceLandmark.NOSE_BASE,
        FaceLandmarkType.MOUTH_LEFT to MlkitFaceLandmark.MOUTH_LEFT,
        FaceLandmarkType.MOUTH_RIGHT to MlkitFaceLandmark.MOUTH_RIGHT,
        FaceLandmarkType.MOUTH_BOTTOM to MlkitFaceLandmark.MOUTH_BOTTOM
      )
      pairs.mapNotNull { (type, mlType) ->
        face.getLandmark(mlType)?.let { lm ->
          FaceLandmark(type, lm.position.x.toDouble(), lm.position.y.toDouble())
        }
      }.toTypedArray()
    } else {
      emptyArray()
    }

    return DetectedFace(
      bounds = FaceBounds(
        x = face.boundingBox.left.toDouble(),
        y = face.boundingBox.top.toDouble(),
        width = face.boundingBox.width().toDouble(),
        height = face.boundingBox.height().toDouble()
      ),
      headEulerAngleY = face.headEulerAngleY.toDouble(),
      headEulerAngleZ = face.headEulerAngleZ.toDouble(),
      leftEyeOpenProbability = (face.leftEyeOpenProbability ?: -1f).toDouble(),
      rightEyeOpenProbability = (face.rightEyeOpenProbability ?: -1f).toDouble(),
      smilingProbability = (face.smilingProbability ?: -1f).toDouble(),
      landmarks = landmarks,
      trackingId = (face.trackingId ?: -1).toDouble()
    )
  }
}
